### PR TITLE
Support Megatron-LM

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Support latest megatron-lm. Can't add UT because megatron is not a library.

**Major Revision**
- DistributedOptimizer support FP8
- DistributedDataParallel support FP8
- clip_grad_norm_fp8
- LinearWithGradAccumulationAndAsyncCommunication with FP8 support

### CODE REVIEW

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

File: .flake8 [hunk 1]
```diff
@@ -5,3 +5,4 @@ inline-quotes = single
 multiline-quotes = double
 docstring-quotes = double
 docstring-convention = google
+per-file-ignores = *.py:D202
```
Comment (guoshzhao) [inline .flake8:8]: Why need this ignore?

File: msamp/megatron/distributed.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""The DistributedDataParallel which supports FP8."""
+
+import math
+import torch
```
Comment (guoshzhao) [inline msamp/megatron/distributed.py:7]: package `torch` is the same with megtraon, both are third party packages, while math is native package. Put torch together with megatron.

File: msamp/megatron/distributed.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                    type_num_elements[dtype] -= param.data.nelement()
+                    param.main_grad = self._grad_buffers[dtype].get(param.data.shape, type_num_elements[dtype])
+                    if dtype not in self._grad_buffer_param_index_map:
+                        self._grad_buffer_param_index_map[dtype] = {}
+                    self._grad_buffer_param_index_map[dtype][param] = (
+                        type_num_elements[dtype],
+                        type_num_elements[dtype] + param.data.nelement(),
+                    )
+
+            self._grad_buffer_num_params = [0 for _ in range(data_parallel_world_size)]
```
Comment (guoshzhao) [inline msamp/megatron/distributed.py:105]: Looks the code added by ourselve has less comments than the original part, could you add more comments?

File: .flake8 (hunk 1 above)
Comment (tocean) [reply .flake8:8]: Because D202 has conflict with yapf formatter.

File: msamp/megatron/distributed.py (hunk 2 above)
Comment (tocean) [reply msamp/megatron/distributed.py]: Good point. Fixed.

File: msamp/megatron/distributed.py (hunk 3 above)
Comment (tocean) [reply msamp/megatron/distributed.py:107]: Sure. Added some comments for our MS-AMP part.

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**